lib/internal/quic/core.js

```diff
@@ -2123,51 +2160,37 @@ class QuicSession extends EventEmitter {
       throw new ERR_INVALID_STATE(
         `${this.constructor.name} is closing`);
     }
+
     const {
       halfOpen,  // Unidirectional or Bidirectional
       highWaterMark,
       defaultEncoding,
     } = validateQuicStreamOptions(options);
 
-    const stream = new QuicStream({
-      highWaterMark,
-      defaultEncoding,
-      readable: !halfOpen
-    }, this);
-
-    state.pendingStreams.add(stream);
-
-    // If early data is being used, we can create the internal QuicStream on the
-    // ready event, that is immediately after the internal QuicSession handle
-    // has been created. Otherwise, we have to wait until the secure event
-    // signaling the completion of the TLS handshake.
-    const makeStream = QuicSession[kMakeStream].bind(this, stream, halfOpen);
-    let deferred = false;
-    if (!this.handshakeComplete) {
-      deferred = true;
-      this.once('secure', makeStream);
-    }
-
-    if (!deferred)
-      makeStream(stream, halfOpen);
+    await this[kHandshakeComplete]();
```

Comment:
Seems a little weird to have a async function returning an async stream? Seems to me this logic should be in `_construct`?

Reply:
I'm not following. Should be in which _construct?

This is resolving once the `QuicStream` is actually available to be used and doing it this way greatly simplifies the state management within the `QuicStream` implementation.

Reply:
To explain further... the challenge here is determining when we can reasonably start writing data outbound data to the `QuicStream`. When we have `QuicSession` instance, the TLS handshake may or may not yet have completed. We cannot actually provision a `QuicStream` handle until it is complete. Because `Writable` has no notion of "Buffer writes until I say we're ready" mechanism outside of cork/uncork, moving the necessary state tracking into the `QuicStream` becomes more difficult. All of that "pending state" machinery is replaced here with a single `await this[kHandshakeComplete]()`, which is deferring the actual creation of the `QuicStream` until we know the handle can actually be provisioned.

Reply:
Streams hade this through _construct now. I’ll make an example once I’m back at my computer tomorrow.

Reply:
@jasnell example:

```js
openStream(options) {
  if (this.destroyed) {
    throw new ERR_INVALID_STATE(
      `${this.constructor.name} is already destroyed`);
  }
  if (this.closing) {
    throw new ERR_INVALID_STATE(
      `${this.constructor.name} is closing`);
  }

  const stream = new QuicStream({
    highWaterMark,
    defaultEncoding,
    readable: !halfOpen
  });

  stream._construct = async function (callback) {
    try {
      const {
        halfOpen,  // Unidirectional or Bidirectional
        highWaterMark,
        defaultEncoding,
      } = validateQuicStreamOptions(options);

      await this[kHandshakeComplete]();

      if (this.destroyed) {
        throw new ERR_INVALID_STATE(
          `${this.constructor.name} is already destroyed`);
      }
      if (this.closing) {
        throw new ERR_INVALID_STATE(
          `${this.constructor.name} is closing`);
      }

      const handle =
        halfOpen ?
          _openUnidirectionalStream(this[kHandle]) :
          _openBidirectionalStream(this[kHandle]);

      if (handle === undefined)
        throw new ERR_OPERATION_FAILED('Unable to create QuicStream');

      // Other code using handle to initialize stream...

      const state = this[kInternalState];
      state.highWaterMark = highWaterMark;
      state.defaultEncoding = defaultEncoding;
      state.session = this;
      state.push_id = null;
      this._readableState.readingMore = true;
      this.on('pause', streamOnPause);
  
      this[kSetHandle](handle);

      callback();
    } catch (err) {
      callback(err);
    }
  }

  return stream;
}
```

Reply:
Wow..lol, I had *completely* forgotten about that :-D ... I do notice that there's no catch handler accounted for on the off change that _construct is an async function. I'll open a separate PR dealing with that case.